SS2-Project/Assets/Starstorm2/Modules/SS2EventCard.cs

```diff
@@ -0,0 +1,110 @@
+﻿using System;
+using UnityEngine;
+using Moonstorm.AddressableAssets;
+using System.Collections.Generic;
+using System.Collections.ObjectModel;
+using RoR2;
+using R2API;
+using UnityEngine.AddressableAssets;
+
+namespace Moonstorm.Starstorm2
+{
+    public class SS2EventCard : EventCard
+    {
+        [Serializable]
+        public class EventVFX
+        {
+            public R2API.DirectorAPI.Stage stageEnum;
+            public string customStageName;
+            public GameObject effectPrefab;
+
+            public GameObject EffectPrefab { get; internal set; }
+        }
+
+        private static List<SS2EventCard> instances = new List<SS2EventCard>();
+        public AddressableItemDef requiredItem;
+        public GameObject fallbackVFX;
+        public EventVFX[] eventVFX = Array.Empty<EventVFX>();
+        
+        public ReadOnlyDictionary<R2API.DirectorAPI.Stage, GameObject> vanillaStageToFXPrefab;
+        public ReadOnlyDictionary<string, GameObject> customStageToFXPrefab;
+
+        private void Awake()
+        {
+            instances?.Add(this);    
+        }
+        private void OnDestroy()
+        {
+            instances?.Remove(this);
+        }
+
+        [SystemInitializer]
+        private static void SystemInit()
+        {
+            EventCatalog.resourceAvailability.CallWhenAvailable(() =>
+            {
+                foreach(var instance in instances)
+                {
+                    instance.CreateVFXDictionary();
+                }
+            });
+        }
+
+        private bool checkRequiredItem() 
+        {
+            if (requiredItem == null)
+            {
+                return true;
+            } else {
+                foreach (var player in PlayerCharacterMasterController.instances)
+                {
+                    var playerBody = player.master.GetBody();
+
+                    if (playerBody != null)
```

Comment:
Use unity's implicit bool operator instead of manually checking if its null.

This should also be a guard clause

Reply:
switched to implicit bool, wdym by guard clause in this case?

Reply:
Guard clauses are used to avoid nesting and produce cleaner readable code

https://youtube.com/shorts/Zmx0Ou5TNJs?feature=share3

Reply:
this is the dumbest name for what us C programmers just call "null checking". OOP Cringe

Reply:
Hmm looking over the code not sure I understand the suggestion here. Theres only one level of nesting and thats just checking if inventory count is sufficient to return. We cant return early if false (since we might not have checked all players yet) , and inventoryCount is local to the scope of the if block

Reply:
in this case instead of using return, you'd use continue, that way it goes to the next element of the list.

Reply:
and also, nesting stinky

Reply:
but we dont need to use continue if we already found the item we can just return early 🤔

Reply:
NESTING STINKYYY REEEEE

fine, you win uwu